commons-impl/src/main/java/com/connexta/transformation/commons/api/impl/Persistable.java

```diff
@@ -0,0 +1,456 @@
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package com.connexta.transformation.commons.api.impl;
+
+import com.connexta.transformation.commons.api.exceptions.InvalidFieldException;
+import com.connexta.transformation.commons.api.exceptions.PersistenceException;
+import com.connexta.transformation.pojo.Pojo;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+/**
+ * This class defines some data that any object to be saved in persistence should have and methods
+ * to manipulate that data.
+ *
+ * @param <P> the type of pojo associated with this persistable
+ */
+public abstract class Persistable<P extends Pojo> {
+  /** A string representing the type of this object. Used when generating exception or logs. */
+  protected final String persistableType;
+
+  private String id;
+
+  /**
+   * Instantiates a new persistable corresponding to a new object that do not yet exist in the
+   * database.
+   *
+   * @param type a string representing the type of this object (used when generating exception or
+   *     logs)
+   */
+  protected Persistable(String type) {
+    this(type, UUID.randomUUID().toString());
+  }
+
+  /**
+   * Instantiates a new persistable corresponding to an object being reloaded from the database in
+   * which case the identifier can either be provided or <code>null</code> can be provided with the
+   * expectation that the identifier will be restored as soon as the subclass calls {@link
+   * #readFrom(Pojo)}.
+   *
+   * <p><i>Note:</i> It is the responsibility of the derived class which calls this constructor with
+   * a <code>null</code> identifier to ensure that the identifier is restored or set before
+   * returning any instances of the object.
+   *
+   * @param persistableType a string representing the type of this object (used when generating
+   *     exception or logs)
+   * @param id the previously generated identifier for this object or <code>null</code> if it is
+   *     expected to be restored later when {@link #readFrom(Pojo)} is called by the subclass
+   */
+  protected Persistable(String persistableType, @Nullable String id) {
+    this.persistableType = persistableType;
+    this.id = id;
+  }
+
+  /**
+   * Gets the identifier of the persistable object.
+   *
+   * @return The ID of the object
+   */
+  public String getId() {
+    return id;
+  }
+
+  /**
+   * Checks if this persistable object contains unknown information.
+   *
+   * @return <code>true</code> if it contains unknown info; <code>false</code> otherwise
+   */
+  public boolean hasUnknowns() {
+    return false;
```

Comment:
✏️ Is it assumed that an extending class will override this method? Should we note that in the javadoc?